### PR TITLE
FEAT: Number theory module

### PR DESCRIPTION
**Pull request by pearsonify**
Added a module that covers the basics of number theory. There are several more functions still to be added and some improvements need to be made to error handling and operation checking.

### CODE REVIEW

File: src/number_theory/numbers.rs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+pub trait Integers:
+    BaseNum + ToPrimitive + FromPrimitive + SampleUniform + Bounded + Send + Sync + Ord + CheckedMul
+{
+}
+
+impl<T> Integers for T where
+    T: BaseNum
+        + ToPrimitive
+        + FromPrimitive
+        + SampleUniform
```
Comment (pearsonify) [inline src/number_theory/numbers.rs:15]: Not all of these traits are needed to cover the set of all integers. However, there are some desired additional functions for other purposes. Such as the ability to generate random integers along a uniform sample number line. The ability to enforce orders and sort. Finally, checked multiplication which is desirable for error checking. However, to actually cover the basics of what an Integer is, BaseNum, Bounded, Send, Sync should be sufficient.

File: src/number_theory/modulus.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+/// Implement the modulo operation
+///
+/// The modulus operator (%) is the remainder operation but not a congruency modulus
+/// See: https://en.wikipedia.org/wiki/Modular_arithmetic for more information
+///
+/// Two functions are implemented, one to find a possible congruency given the modulus, another to check if a and b are congruent given the modulus
+pub trait Modulus<T> {
+    fn modulus(self, b: T) -> T;
+    // fn is_congruent_modulus(self, b: T) -> bool;
```
Comment (pearsonify) [inline src/number_theory/modulus.rs:13]: Two problems:
1. Not implemented or in TODO
2. Wrong number of arguments, a is congruent to b mod c

File: src/number_theory/parity.rs [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    fn parity(self) -> String;
+    fn is_odd(self) -> bool;
+    fn is_even(self) -> bool;
+}
+
+impl<T> Parity<T> for T
+where
+    T: Integers,
+{
+    fn parity(self) -> String {
```
Comment (pearsonify) [inline src/number_theory/parity.rs:15]: Is this ever useful? Probably not. Is anything in this trait non-obvious? No.

File: src/number_theory/gcd.rs [hunk 4]
```diff
@@ -0,0 +1,82 @@
+//! Greatest Common Divisor Trait
+use crate::number_theory::numbers::Integers;
+/// Greatest Common Divisor
+///
+/// This trait contains multiple function implementations of algorithms for finding the greatest common divisor
+pub trait GCD<T> {
+    fn gcd(self, other: T) -> T;
+    fn gcd_subtraction(self, other: T) -> T;
```
Comment (pearsonify) [inline src/number_theory/gcd.rs:8]: Checked subtraction or wrapping subtraction?

File: src/number_theory/factorial.rs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+use crate::number_theory::numbers::Integers;
+use anyhow::{anyhow, Result};
+
+/// Implement the factorial operation
+pub trait Factorial<T = Self> {
+    fn factorial(&self) -> Result<T>;
+}
+
+impl<T: Integers> Factorial<T> for T {
+    fn factorial(&self) -> Result<T> {
```
Comment (pearsonify) [inline src/number_theory/factorial.rs:11]: Does checked make more sense than wrapping here? Yes, rather than shifting bits, this operation should fail on overflow for the type.

File: src/number_theory/continued_fraction.rs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        let mut temp = T::zero();
+        let mut a = self;
+        let mut b = other;
+        loop {
+            // Find the integer portion of the fraction
+            i = a / b;
+            fraction.push(i);
+
+            // Subtract the integer portion of the fraction
+            a = a - (b * i);
```
Comment (pearsonify) [inline src/number_theory/continued_fraction.rs:28]: Checked operations?

File: src/number_theory/README.md [hunk 7]
```diff
@@ -0,0 +1,17 @@
+# Number Theory Module
+
+#### Overview 
+
+##### Why Number Theory?
```
Comment (pearsonify) [inline src/number_theory/README.md]: Out of place now that this was transferred out of the cryptography module.

File: src/number_theory/numbers.rs (hunk 1 above)
Comment (pearsonify) [reply src/number_theory/numbers.rs:15]: The additional traits belong here due to the dependency tree existent between the different modules being worked on (to be committed to separate branches when ready).

File: src/number_theory/parity.rs (hunk 3 above)
Comment (pearsonify) [reply src/number_theory/parity.rs:15]: But it exists.

File: src/number_theory/README.md (hunk 7 above)
Comment (pearsonify) [reply src/number_theory/README.md]: True. This will be rewritten.

**Comment by pearsonify** [comment]
There may need to be some way to check if unsigned for the trait objects. Maybe separating Integers into Integers, Positive Integers, and Negative Integers (where both positive and negative are inclusive of 0). Some additions such as the extended euclidean algorithm for computing multiplicative inverses do not make sense when restricted to positive integers alone.

The immediate issues here are clear:
 - Separate trait objects would be required for positive and negative integers. Any trait using the Integers trait would fail if the algorithm is invalid for positive integers only since it would still be possible to apply that trait to unsigned types.
 - The traits would be extremely redundant. The bounding of the number line would be the only change. This is not a property of the types themselves. They would all still be trait bound by the Bounded trait, but the number lines would be bounded at different positions.

In conclusion, this suggested methodology does not make sense. I think potentially this is a case where the implementation of an algorithm that is only valid for all integers may be best to do as a macro or potentially just give the user a warning that it is only valid when all integers are available.
